### PR TITLE
type.convert arg

### LINKED ISSUE #28: Offer 'type.convert' in interface?
Finally giving {nc} a try, it's quite handy!

It took me a while to find out `capture_first_vec()` offers a way to do type conversion:

> Each function must be un-named, and is used to convert the previous capture `group`.

This is nice, but maybe "too much" flexibility. I think `utils::type.convert()` would offer a much simpler interface here, instead of 4 different `as.integer` entries, I could just pass `type.convert=TRUE` and let R figure out the appropriate type for me. That's also the approach `data.table::tstrsplit()` takes which would make it more familiar as well.

### PR DESCRIPTION
**Pull request by tdhock**
Closes #28 

TODO add tests

### CODE REVIEW

File: tests/testthat/test-CRAN-vec.R [hunk 1]
```diff
@@ -101,6 +101,70 @@ test_engines("capture_first_vec returns data.table with int columns", {
...
+})
+
+test_engines("capture_all_str(type.convert=TRUE) returns two int columns", {
+  computed <- capture_all_str(
+    "chr1:2-3,000 chr4:5-6,000",
+    chrom="chr.*?",
+    ":",
+    chromStart=".*?",
+    "-",
+    chromEnd="[0-9,]*", keep.digits,
```
Comment (MichaelChirico) [inline tests/testthat/test-CRAN-vec.R]: nice, really like that we get this "default with override" behavior "for free"

**Review by MichaelChirico — COMMENTED** [review]
LGTM and it's definitely what I had in mind.

Glancing at the PR, I wonder if it would be a better API to instead offer "default.coercion=identity", then I can use default.coercion=type.convert but also we allow even more flexibility.

put another way, before we have (1) only custom column-wise conversion allowed, here we have (2) custom column-wise conversion, possibly defaulting to type.convert(). But maybe we could get (3) custom column-wise conversion backing up to custom default conversion.

**Comment by tdhock** [comment]
thanks for sharing your additional ideas. That makes sense to me but I kept the argument name as `type.convert` instead of your suggested `default.coercion` because "coercion" was not mentioned anywhere in the code/docs, but "convert" and "conversion" were.

**Comment by tdhock** [comment]
also I noticed that I had to specify type.convert(x,as.is=TRUE) to get character output and avoid a warning, so I kept the type.convert=TRUE alias for that as a shortcut.

**Comment by tdhock** [comment]
do you have a compelling example in mind that we could add to a man page or vignette?

For the SLURM sacct example (v1-capture-first vignette) we convert everything to integer but the existing example handles that by integrating the conversion function with the pattern:
```r
(sacct.df <- data.frame(
  Elapsed = c(
    "07:04:42", "07:04:42", "07:04:49",
    "00:00:00", "00:00:00"),
  JobID=c(
    "13937810_25",
    "13937810_25.batch",
    "13937810_25.extern",
    "14022192_[1-3]",
    "14022204_[4]"),
  stringsAsFactors=FALSE))
int.pattern <- list("[0-9]+", as.integer)
range.pattern <- list(
  "\\[",
  task1=int.pattern,
  list(
    "-",#begin optional end of range.
    taskN=int.pattern
  ), "?", #end is optional.
  "\\]")
nc::capture_first_df(sacct.df, JobID=range.pattern, nomatch.error=FALSE)
```
gives
```r
> nc::capture_first_df(sacct.df, JobID=range.pattern, nomatch.error=FALSE)
    Elapsed              JobID task1 taskN
     <char>             <char> <int> <int>
1: 07:04:42        13937810_25    NA    NA
2: 07:04:42  13937810_25.batch    NA    NA
3: 07:04:49 13937810_25.extern    NA    NA
4: 00:00:00     14022192_[1-3]     1     3
5: 00:00:00       14022204_[4]     4    NA
```

**Comment by MichaelChirico** [comment]
> do you have a compelling example in mind that we could add to a man page or vignette?

I came to this via Advent of Code, specifically this year's Day 14 problem:

https://github.com/MichaelChirico/advent-of-code/blob/dc272d9c342fa4cb35e8519f5f41cd8d3e17b54d/2024/14.R#L1-L14

https://adventofcode.com/2024/day/14

In case that requires log-in to view, here's a sample file:

https://github.com/MichaelChirico/advent-of-code/blob/main/2024/input-data/14-test

The Advent of Code problems generally provide good motivating examples for {nc}, as the problem input often takes the form of structured text files from which you need to extract data in a friendlier format.

**Comment by tdhock** [comment]
thanks for sharing the example.

**Pull request merged**